### PR TITLE
(Refactor) Document and clean single invocation partition.

### PR DESCRIPTION
**Pull request by ajreynol**
This pull request documents and cleans one of the key utilities for single invocation synthesis (SingleInvocationPartition), and creates a new file dedicated to this utility.

Most of the changes are to deal with the encapsulation of the data members of the utility.

NOTE:  this is refactor commit so the file should not be reviewed:
src/theory/quantifiers/single_inv_partition.cpp
since it was moved from src/theory/quantifiers/ce_guided_single_inv.cpp

### CODE REVIEW

File: src/theory/quantifiers/single_inv_partition.h [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+ * see Example 5 of Reynolds et al. SYNT 2017.
+ *
+ */
+class SingleInvocationPartition
+{
+ public:
+  SingleInvocationPartition() : d_has_input_funcs(false) {}
+  ~SingleInvocationPartition() {}
+  /** initialize this partition for formula n, with input functions funcs
+   *
```
Comment (HanielB) [inline src/theory/quantifiers/single_inv_partition.h]: newlines before docstrings

File: src/theory/quantifiers/single_inv_partition.h [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+   */
+  Node getNonSingleInvocation() { return getConjunct(1); }
+  /** get full specification
+   *
+   * This returns getSingleInvocation() * sigma ^ getNonSingleInvocation(),
+   * which is equivalent to the processed formula, where sigma is the
+   * substitution described above.
+   */
+  Node getFullSpecification() { return getConjunct(2); }
+  /** get first order variable for input function f
```
Comment (HanielB) [inline src/theory/quantifiers/single_inv_partition.h]: Make docstrings consistent. The above have a newline between the brief and the description.

File: src/theory/quantifiers/single_inv_partition.h [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+   * This corresponds to the variable that we used when anti-skolemized
```
Comment (HanielB) [inline src/theory/quantifiers/single_inv_partition.h]: skolemized --> skolemizing

File: src/theory/quantifiers/single_inv_partition.h [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+  std::vector<Node> d_func_vars;
+  /** the arguments that we based the anti-skolemization on.
+   * In (EX1), this is the list { x, y }.
+   */
+  std::vector<Node> d_si_vars;
+  /** every free variable of conjuncts[2] */
+  std::vector<Node> d_all_vars;
+  /** map from functions to first-order variables that anti-skolemized them */
+  std::map<Node, Node> d_func_fo_var;
+  /** map from first-order variables to the function anti-skolemized */
```
Comment (HanielB) [inline src/theory/quantifiers/single_inv_partition.h]: put adjective first

File: src/theory/quantifiers/single_inv_partition.h [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+  /** did we initialize this class with input functions? */
+  bool d_has_input_funcs;
+  /** the input functions we initialized this class with */
+  std::vector<Node> d_input_funcs;
+  /** all input functions */
+  std::vector<Node> d_all_funcs;
+
+  /** infer the argument types of uninterpreted function applications
+   *
+   * If this method returns true, then the contents of typs contains
```
Comment (HanielB) [inline src/theory/quantifiers/single_inv_partition.h]: contains --> contain

File: src/smt/smt_engine.cpp [hunk 6]
```diff
@@ -4628,13 +4629,16 @@ Result SmtEngine::checkSynth(const Expr& e) throw(Exception) {
...
                            << std::endl;
         // partition variables
+        std::vector<Node> all_vars;
+        sip.getAllVariables(all_vars);
+        std::vector<Node> si_vars;
+        sip.getSingleInvocationVariables(si_vars);
         std::vector<Node> qe_vars;
         std::vector<Node> nqe_vars;
-        for (unsigned i = 0; i < sip.d_all_vars.size(); i++)
+        for (unsigned i = 0; i < all_vars.size(); i++)
```
Comment (aniemetz) [inline src/smt/smt_engine.cpp]: Do not call size() in each iteration.

File: src/smt/smt_engine.cpp [hunk 7]
```diff
@@ -4657,18 +4661,21 @@ Result SmtEngine::checkSynth(const Expr& e) throw(Exception) {
           Trace("smt-synth") << "  subs : " << nqe_vars[i] << " -> " << k
                              << std::endl;
         }
-        for (std::map<Node, bool>::iterator it = sip.d_funcs.begin();
-             it != sip.d_funcs.end();
-             ++it)
+        std::vector<Node> funcs;
+        sip.getFunctions(funcs);
+        for (unsigned i = 0; i < funcs.size(); i++)
```
Comment (aniemetz) [inline src/smt/smt_engine.cpp]: Ditto

File: src/smt/smt_engine.cpp [hunk 8]
```diff
@@ -4657,18 +4661,21 @@ Result SmtEngine::checkSynth(const Expr& e) throw(Exception) {
...
         {
-          orig.push_back(sip.d_func_inv[it->first]);
+          Node f = funcs[i];
+          Node fi = sip.getFunctionInvocationFor(f);
+          Node fv = sip.getFirstOrderVariableForFunction(f);
+          Assert(!fi.isNull());
+          orig.push_back(fi);
           Node k = NodeManager::currentNM()->mkSkolem(
```
Comment (aniemetz) [inline src/smt/smt_engine.cpp]: Do not call currentNM() in each iteration of the loop.

File: src/theory/quantifiers/anti_skolem.cpp [hunk 9]
```diff
@@ -114,7 +114,12 @@ void QuantAntiSkolem::check( Theory::Effort e, unsigned quant_e ) {
         }
         if( success ){
           //sort the argument variables
-          d_ask_types[q].insert( d_ask_types[q].end(), d_quant_sip[q].d_arg_types.begin(), d_quant_sip[q].d_arg_types.end() );
+          std::vector<Node> sivars;
+          d_quant_sip[q].getSingleInvocationVariables(sivars);
+          for (unsigned j = 0; j < sivars.size(); j++)
```
Comment (aniemetz) [inline src/theory/quantifiers/anti_skolem.cpp]: Ditto

File: src/theory/quantifiers/anti_skolem.cpp [hunk 10]
```diff
@@ -114,7 +114,12 @@ void QuantAntiSkolem::check( Theory::Effort e, unsigned quant_e ) {
...
+          {
+            d_ask_types[q].push_back(sivars[j].getType());
+          }
           std::map< TypeNode, std::vector< unsigned > > indices;
           for( unsigned j=0; j<d_ask_types[q].size(); j++ ){
```
Comment (aniemetz) [inline src/theory/quantifiers/anti_skolem.cpp]: Ditto

File: src/theory/quantifiers/anti_skolem.cpp [hunk 11]
```diff
@@ -163,8 +168,11 @@ bool QuantAntiSkolem::sendAntiSkolemizeLemma( std::vector< Node >& quants, bool
       for( unsigned i=0; i<quants.size(); i++ ){
         Node q = quants[i];
         std::vector< int > eqcs;
-        for( std::map< Node, bool >::iterator it = d_quant_sip[q].d_funcs.begin(); it != d_quant_sip[q].d_funcs.end(); ++it ){
-          Node f = it->first;
+        std::vector<Node> funcs;
+        d_quant_sip[q].getFunctions(funcs);
+        for (unsigned j = 0; j < funcs.size(); j++)
```
Comment (aniemetz) [inline src/theory/quantifiers/anti_skolem.cpp]: Ditto

File: src/theory/quantifiers/anti_skolem.cpp [hunk 12]
```diff
@@ -235,15 +243,22 @@ bool QuantAntiSkolem::sendAntiSkolemizeLemma( std::vector< Node >& quants, bool
       std::vector< Node > subs_rhs;
       //get outer variable substitution
       Assert( d_ask_types_index[q].size()==d_ask_types[q].size() );
+      std::vector<Node> sivars;
+      d_quant_sip[q].getSingleInvocationVariables(sivars);
       for( unsigned j=0; j<d_ask_types_index[q].size(); j++ ){
```
Comment (aniemetz) [inline src/theory/quantifiers/anti_skolem.cpp]: Ditto

File: src/theory/quantifiers/anti_skolem.cpp [hunk 13]
```diff
@@ -235,15 +243,22 @@ bool QuantAntiSkolem::sendAntiSkolemizeLemma( std::vector< Node >& quants, bool
...
+        subs_lhs.push_back(sivars[d_ask_types_index[q][j]]);
         subs_rhs.push_back( outer_vars[j] );
       }
       //get function substitution
-      for( std::map< Node, bool >::iterator it = d_quant_sip[q].d_funcs.begin(); it != d_quant_sip[q].d_funcs.end(); ++it ){
-        Node f = it->first;
-        Node fv = d_quant_sip[q].d_func_fo_var[it->first];
+      std::vector<Node> funcs;
+      d_quant_sip[q].getFunctions(funcs);
+      for (unsigned i = 0; i < funcs.size(); i++)
```
Comment (aniemetz) [inline src/theory/quantifiers/anti_skolem.cpp]: Ditto

File: src/theory/quantifiers/ce_guided_single_inv.cpp [hunk 14]
```diff
@@ -188,17 +177,25 @@ void CegConjectureSingleInv::initialize( Node q ) {
...
+                                             prog_templ_vars.end());
             Trace("cegqi-inv") << "      invariant : " << invariant << std::endl;
             
             // store simplified version of quantified formula
-            d_simp_quant = d_sip->d_conjuncts[2].size()==1 ? d_sip->d_conjuncts[2][0] : NodeManager::currentNM()->mkNode( AND, d_sip->d_conjuncts[2] );
+            d_simp_quant = d_sip->getFullSpecification();
             std::vector< Node > new_bv;
-            for( unsigned j=0; j<d_sip->d_si_vars.size(); j++ ){
-              new_bv.push_back( NodeManager::currentNM()->mkBoundVar( d_sip->d_si_vars[j].getType() ) );
+            for (unsigned j = 0; j < sivars.size(); j++)
```
Comment (aniemetz) [inline src/theory/quantifiers/ce_guided_single_inv.cpp]: Ditto

File: src/theory/quantifiers/ce_guided_single_inv.cpp [hunk 15]
```diff
@@ -188,17 +177,25 @@ void CegConjectureSingleInv::initialize( Node q ) {
...
+            {
+              new_bv.push_back(
+                  NodeManager::currentNM()->mkBoundVar(sivars[j].getType()));
```
Comment (aniemetz) [inline src/theory/quantifiers/ce_guided_single_inv.cpp]: Ditto

File: src/theory/quantifiers/ce_guided_single_inv.cpp [hunk 16]
```diff
@@ -283,17 +280,26 @@ void CegConjectureSingleInv::finishInit( bool syntaxRestricted, bool hasItes ) {
...
+    {
+      Node pbvl = NodeManager::currentNM()->mkNode(BOUND_VAR_LIST, func_vars);
       d_single_inv = NodeManager::currentNM()->mkNode( FORALL, pbvl, d_single_inv );
     }
     //now, introduce the skolems
-    for( unsigned i=0; i<d_sip->d_si_vars.size(); i++ ){
-      Node v = NodeManager::currentNM()->mkSkolem( "a", d_sip->d_si_vars[i].getType(), "single invocation arg" );
+    std::vector<Node> sivars;
+    d_sip->getSingleInvocationVariables(sivars);
+    for (unsigned i = 0; i < sivars.size(); i++)
```
Comment (aniemetz) [inline src/theory/quantifiers/ce_guided_single_inv.cpp]: Ditto

File: src/theory/quantifiers/ce_guided_single_inv.cpp [hunk 17]
```diff
@@ -283,17 +280,26 @@ void CegConjectureSingleInv::finishInit( bool syntaxRestricted, bool hasItes ) {
...
+    {
+      Node v = NodeManager::currentNM()->mkSkolem(
```
Comment (aniemetz) [inline src/theory/quantifiers/ce_guided_single_inv.cpp]: Ditto

File: src/theory/quantifiers/anti_skolem.cpp [hunk 18]
```diff
@@ -172,7 +173,7 @@ bool QuantAntiSkolem::sendAntiSkolemizeLemma( std::vector< Node >& quants, bool
         std::vector< int > eqcs;
         std::vector<Node> funcs;
         d_quant_sip[q].getFunctions(funcs);
-        for (unsigned j = 0; j < funcs.size(); j++)
+        for (unsigned j = 0, size = funcs.size(); j < size; j++)
```
Comment (aniemetz) [inline src/theory/quantifiers/anti_skolem.cpp]: for-each?

File: src/theory/quantifiers/anti_skolem.cpp [hunk 19]
```diff
@@ -257,7 +260,7 @@ bool QuantAntiSkolem::sendAntiSkolemizeLemma( std::vector< Node >& quants, bool
       //get function substitution
       std::vector<Node> funcs;
       d_quant_sip[q].getFunctions(funcs);
-      for (unsigned i = 0; i < funcs.size(); i++)
+      for (unsigned i = 0, size = funcs.size(); i < size; i++)
```
Comment (aniemetz) [inline src/theory/quantifiers/anti_skolem.cpp]: for-each?

**Pull request merged**